### DISCUSSION TITLE
macOS 上设置了 `ExtensionInstallAllowlist` 仍然无法安装 crx

### DISCUSSION THREAD

**Discussion by aur3l14no** [ronggang/PT-Plugin-Plus#1066] [Q&A] [answered]
``` bash
~ ❯ defaults read com.google.Chrome
{
    ExtensionInstallAllowlist =     (
        dmmjlmbkigbgpnjfiimhlnbnmppjhpea
    );
}
```

但仍然提示 This extension is not listed in the Chrome Web Store and may have been added without your knowledge.

**Comment by aur3l14no** [comment, marked as answer]
原因是 defaults 写入的 policy 并非 mandatory... 可通过写入 `/Library/Managed Preferences/com.google.Chrome.plist` 或安装下面的这个 描述文件来设置成 mandatory

保存为 `foobar.mobileconfig`, 安装，在 Preference -> Profile 中激活即可

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>PayloadContent</key>
    <array>
        <dict>
            <key>PayloadContent</key>
            <dict>
                <key>com.google.Chrome</key>
                <dict>
                    <key>Forced</key>
                    <array>
                        <dict>
                            <key>mcx_preference_settings</key>
                            <dict>
                                <key>ExtensionInstallAllowlist</key>
                                <array>
                                    <string>dmmjlmbkigbgpnjfiimhlnbnmppjhpea</string>
                                </array>
                            </dict>
                        </dict>
                    </array>
                </dict>
            </dict>
            <key>PayloadEnabled</key>
            <true/>
            <key>PayloadIdentifier</key>
            <string>MCXToProfile.7e2bec75-299e-44ff-b405-628007abffff.alacarte.customsettings.bdac4880-d25f-4cdd-8472-05473f005e7e</string>
            <key>PayloadType</key>
            <string>com.apple.ManagedClient.preferences</string>
            <key>PayloadUUID</key>
            <string>bdac4880-d25f-4cdd-8472-05473f005e7e</string>
            <key>PayloadVersion</key>
            <integer>1</integer>
        </dict>
    </array>
    <key>PayloadDescription</key>
    <string>Included custom settings:
com.google.Chrome
</string>
    <key>PayloadDisplayName</key>
    <string>MCXToProfile: com.google.Chrome</string>
    <key>PayloadIdentifier</key>
    <string>com.google.Chrome</string>
    <key>PayloadOrganization</key>
    <string></string>
    <key>PayloadRemovalDisallowed</key>
    <true/>
    <key>PayloadScope</key>
    <string>System</string>
    <key>PayloadType</key>
    <string>Configuration</string>
    <key>PayloadUUID</key>
    <string>7e2bec75-299e-44ff-b405-628007abffff</string>
    <key>PayloadVersion</key>
    <integer>1</integer>
</dict>
</plist>
```


相关链接
- https://t.me/ptpluginplus/110
- https://t.me/ptpluginplus/115
- https://support.google.com/chrome/a/answer/7517624?hl=en
- https://stackoverflow.com/questions/38206916/mandatory-chrome-policies-on-mac.

**Reply by KesaubeEire** [reply]
非常感谢, 但是写的有点不太清楚, 找了半天 Preferens -> Profile 不知道是啥, 最后才发现是 MacOS 的"系统偏好设置" 而不是 Chrome 的, 这里写个给后来人的清楚步骤
1. 先把 .crx 扔进 chrome 跳出 无法安装 的红字
2. 找个地方新建一个文件 `foobar.mobileconfig` , foobar 就是个名字随便改, 就是后缀名要对, 把上面代码放进文件
3. 双击安装, 右上角应该会跳显示 "描述文件安装"
4. 打开 MacOS 的左上角 "系统偏好设置", MacOS 12.1 会多一个 "描述文件", 别的版本多找找在哪, 点进去安装
5. Chrome 右上角三个点点开菜单最下面显示 "由贵组织管理"
6. 现在会跳出来安装好的提示了

**Reply by claresongs** [reply]
12.6系统提示无法打开描述文件:
此描述文件缺少必须的信息，或包含格式无效的信息。（NScocoaErrorDomain:3840）
怎么破。。

**Reply by ted423** [reply]
> 12.6系统提示无法打开描述文件: 此描述文件缺少必须的信息，或包含格式无效的信息。（NScocoaErrorDomain:3840） 怎么破。。

查了下，这个似乎是文件文本格式的问题，是否经过了别的系统创建这个文件

**Reply by claresongs** [reply]
> > 12.6系统提示无法打开描述文件: 此描述文件缺少必须的信息，或包含格式无效的信息。（NScocoaErrorDomain:3840） 怎么破。。
> 
> 查了下，这个似乎是文件文本格式的问题，是否经过了别的系统创建这个文件

感谢提供思路,已解决.这里很有意思的一点是必须直接新建后缀为mobileconfig的描述文件,如果新建txt后改后缀名就会出现上面的问题...
第一次用macbook很不熟悉,现在终于可以随时随地打开ptpp了,感谢大佬~

**Reply by KesaubeEire** [reply]
> 非常感谢, 但是写的有点不太清楚, 找了半天 Preferens -> Profile 不知道是啥, 最后才发现是 MacOS 的"系统偏好设置" 而不是 Chrome 的, 这里写个给后来人的清楚步骤
> 
> 1. 先把 .crx 扔进 chrome 跳出 无法安装 的红字
> 2. 找个地方新建一个文件 `foobar.mobileconfig` , foobar 就是个名字随便改, 就是后缀名要对, 把上面代码放进文件
> 3. 双击安装, 右上角应该会跳显示 "描述文件安装"
> 4. 打开 MacOS 的左上角 "系统偏好设置", MacOS 12.1 会多一个 "描述文件", 别的版本多找找在哪, 点进去安装
> 5. Chrome 右上角三个点点开菜单最下面显示 "由贵组织管理"
> 6. 现在会跳出来安装好的提示了

1. 时隔几个月回来更新, 为了升级 PTPP 重新走了下流程, 我这个方案直接安装基本上没啥问题的, 但是版本更新的时候一开始扔进去 .crx 报一个奇怪的错误, 反正解决方案是把 PTPP 先停掉, 再拖进去 .crx 就行了...
2. Mac 升级到了 13.0, 那个描述文件在 MacOS的 系统设置 -> 隐私与安全性 -> (最下面的) 其他 -> 描述文件, Mac 这老六又变位置重找了一遍

**Reply by waght** [reply]
对了，我最后补充一句，要还要重启浏览器才可以。

**Reply by duleigiser** [reply]
位置现在设备管理内

**Comment by zgpinguo** [comment]
mobileconfig安装过程中，需要手动打开 系统偏好设置 ，选择描述文件安装。有没有更智能些，直接安装成功。
例如 [aur3l14no](https://github.com/aur3l14no) 所说，可通过写入 /Library/Managed Preferences/com.google.Chrome.plist，具体该怎么操作。
或者说通过applescript完成的？

**Comment by krio6839** [comment]
这是Google浏览器的解决办法，Edge浏览器有没有

**Comment by SteinX** [comment]
发一个 edge 的给需要的朋友:

[com.microsoft.Edge.mobileconfig.zip](https://github.com/pt-plugins/PT-Plugin-Plus/files/10529090/com.microsoft.Edge.mobileconfig.zip)

**Comment by limoHey** [comment]
如果您的文本编辑器中没有“Unicode (UTF-8)”选项，则可以尝试使用以下步骤保存Mobileconfig文件：

在Mac上打开文本编辑器（TextEdit）。

将该代码复制并粘贴到文本编辑器中。

在文本编辑器中，单击“Format”菜单，然后选择“Make Plain Text”。

在弹出窗口中单击“OK”。

然后单击“File”菜单，选择“Save As”。

在“Save As”对话框中，选择“Desktop”文件夹作为保存位置。

在“Name”字段中输入文件名，并在文件名后添加“.mobileconfig”扩展名。例如：“ChromeSettings.mobileconfig”。

在“File Format”选项中，选择“UTF-8”选项，并单击“Save”按钮。

您现在可以使用Apple Configurator或移动设备管理（MDM）将Mobileconfig文件部署到iOS设备上。

请注意，Mobileconfig文件只能通过移动设备管理（MDM）或Apple Configurator部署到iOS设备上。如果您没有MDM或Apple Configurator，则无法部署Mobileconfig文件。

**Comment by yalunkong** [comment]
现在的问题是，chrome在mac下经常遇到chrome helper导致的cpu占用率爆表，改用chome内核的brave之后，菜单里找不到“由贵组织管理”，chome还是正常的。

**Reply by ted423** [reply]
https://github.com/brave/brave-browser/issues/2457

**Reply by yalunkong** [reply]
谢谢，解决了
我用的是mac，查到可用的解决办法有两种：
1、tar -xf解压ptp的crx文件，然后在brave的扩展程序管理里选择“加载已解压的扩展程序”。我用这个办法，成功装上了。之前直接加载crx也能装，但激活的开关是灰的，不能点。
2、去brave#2457找esjarc的回复
windows似乎可以通过改注册表实现

**Comment by iwinux** [comment]
这是唯一的解决方法了吗？不想搞一个 MDM Profile

**Comment by cikichen** [comment]
crx能安装，但是安装成功后启用按钮是灰色的。
